### PR TITLE
Use protobuf on k8s client

### PR DESCRIPTION
**Pull request by aanm**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7414)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by eloycoto** [comment]
Awesome change, just a thing, shouldn't we update this?
https://github.com/cilium/cilium/blob/8827a1ef79d4d149cce358315de60ae6d7cb57dc/pkg/policy/groups/helpers.go#L102-L120

Also, the watchers will use Rest instead of Protobuf no?

File: pkg/k8s/annotate.go [hunk 1]
```diff
@@ -35,6 +36,11 @@ type K8sClient struct {
 	kubernetes.Interface
 }
 
+// K8sCiliumClient is a wrapper around cilium.Interface.
```
Comment (nebril) [inline pkg/k8s/annotate.go]: not sure, but `s/cilium.Interface/clientset.Interface` ?

**Review by nebril — CHANGES_REQUESTED** [review]
One typo, but feel free to disregard review if I am wrong on that account

**Comment by aanm** [comment]
> Awesome change, just a thing, shouldn't we update this?
> [cilium/pkg/policy/groups/helpers.go](https://github.com/cilium/cilium/blob/8827a1ef79d4d149cce358315de60ae6d7cb57dc/pkg/policy/groups/helpers.go#L102-L120)
> 
> Lines 102 to 120 in [8827a1e](/cilium/cilium/commit/8827a1ef79d4d149cce358315de60ae6d7cb57dc)
> 
>  func getK8sClient() (*clientset.Clientset, error) { 
>  	k8sMutex.Lock() 
>  	defer k8sMutex.Unlock() 
>  	if globalK8sClient != nil { 
>  		return globalK8sClient, nil 
>  	} 
>   
>  	restConfig, err := k8s.CreateConfig() 
>  	if err != nil { 
>  		return nil, fmt.Errorf("Unable to create rest configuration: %s", err) 
>  	} 
>  	k8sClient, err := clientset.NewForConfig(restConfig) 
>  	if err != nil { 
>  		return nil, fmt.Errorf("Unable to create Kubernetes configuration: %s", err) 
>   
>  	} 
>  	globalK8sClient = k8sClient 
>  	return globalK8sClient, nil 
>  } 
> Also, the watchers will use Rest instead of Protobuf no?

@eloycoto which watchers?

**Comment by aanm** [comment]
previous k8s test results https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-K8s/1814/

**Comment by aanm** [comment]
test-me-please

**Comment by eloycoto** [comment]
@aanm these lines: 

https://github.com/cilium/cilium/blob/8827a1ef79d4d149cce358315de60ae6d7cb57dc/daemon/k8s_watcher.go#L333-L347

**Comment by ianvernon** [comment]
@aanm can you provide more context here? What are the gains of using protobuf for the client?

**Comment by aanm** [comment]
tests failed were related with FQDN https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/10781/

a commit was added for benchmark tests which are not run on any CI so merging without re-running the tests again

**Comment by aanm** [comment]
> @aanm these lines:
> 
> [cilium/daemon/k8s_watcher.go](https://github.com/cilium/cilium/blob/8827a1ef79d4d149cce358315de60ae6d7cb57dc/daemon/k8s_watcher.go#L333-L347)
> 
> Lines 333 to 347 in [8827a1e](/cilium/cilium/commit/8827a1ef79d4d149cce358315de60ae6d7cb57dc)
> 
>  func (d *Daemon) EnableK8sWatcher(queueSize uint) error { 
>  	if !k8s.IsEnabled() { 
>  		log.Debug("Not enabling k8s event listener because k8s is not enabled") 
>  		return nil 
>  	} 
>  	log.Info("Enabling k8s event listener") 
>   
>  	restConfig, err := k8s.CreateConfig() 
>  	if err != nil { 
>  		return fmt.Errorf("Unable to create rest configuration: %s", err) 
>  	} 
>   
>  	apiextensionsclientset, err := apiextensionsclient.NewForConfig(restConfig) 
>  	if err != nil { 
>  		return fmt.Errorf("Unable to create rest configuration for k8s CRD: %s", err)

@eloycoto that `apiextensionsclientset` is only to create a CRD so it's all good

> @aanm can you provide more context here? What are the gains of using protobuf for the client?

@ianvernon see https://github.com/cilium/cilium/pull/7414/commits/7274b0e4d040c43f526fffe9fb9170c6eea6a034

**Comment by tgraf** [comment]
Providing some additional context here (cc @ianvernon):

During scale testing with 1k nodes, we have identified increased memory and CPU consumption as well as considerable network bandwidth used. The CPU and memory consumption is primarily due to JSON decoding. Use of protobuf will help here.

The network bandwidth usage is increased because JSON takes up more space than optimized protofbuf. Andre has posted the savings as measured with a benchmark.

**Comment by ianvernon** [comment]
Great, thanks for the explanation ! Change looks great :D

**Pull request merged**